src/ledger/LedgerManagerImpl.cpp

```diff
@@ -881,37 +909,43 @@ LedgerManagerImpl::checkDbState()
 }
 
 void
-LedgerManagerImpl::advanceLedgerPointers()
+LedgerManagerImpl::advanceLedgerPointers(std::shared_ptr<LedgerHeaderReference> header)
 {
+    auto ledgerHash = sha256(xdr::xdr_to_opaque(header->header()));
     CLOG(DEBUG, "Ledger") << "Advancing LCL: "
                           << ledgerAbbrev(mLastClosedLedger) << " -> "
-                          << ledgerAbbrev(mCurrentLedger);
+                          << ledgerAbbrev(header->header(), ledgerHash);
 
-    mLastClosedLedger.hash = mCurrentLedger->getHash();
-    mLastClosedLedger.header = mCurrentLedger->mHeader;
+    mLastClosedLedger.hash = ledgerHash;
+    mLastClosedLedger.header = header->header();
 
-    mCurrentLedger = make_shared<LedgerHeaderFrame>(mLastClosedLedger);
+    ++header->header().ledgerSeq;
+    header->header().previousLedgerHash = ledgerHash;
     CLOG(DEBUG, "Ledger") << "New current ledger: seq="
-                          << mCurrentLedger->mHeader.ledgerSeq;
+                          << header->header().ledgerSeq;
 }
 
 void
 LedgerManagerImpl::processFeesSeqNums(std::vector<TransactionFramePtr>& txs,
-                                      LedgerDelta& delta)
+                                      LedgerState& ls)
 {
     CLOG(DEBUG, "Ledger") << "processing fees and sequence numbers";
     int index = 0;
     try
     {
-        soci::transaction sqlTx(mApp.getDatabase().getSession());
+        // TODO(jonjove): Need a LedgerState here?
         for (auto tx : txs)
         {
-            LedgerDelta thisTxDelta(delta);
-            tx->processFeeSeqNum(thisTxDelta, *this);
-            tx->storeTransactionFee(*this, thisTxDelta.getChanges(), ++index);
-            thisTxDelta.commit();
+            LedgerState lsInner(ls);
+            auto header = lsInner.loadHeader();
+            uint32_t ledgerSeq = header->header().ledgerSeq;
+            header->invalidate();
+
+            tx->processFeeSeqNum(lsInner);
+            tx->storeTransactionFee(mApp.getDatabase(), ledgerSeq, lsInner.getChanges(), ++index);
+            lsInner.updateLastModified();
```

Comment:
why is this `updateLastModified` call here?
Did you add it because of the one missing in `processFeeSeqNum`?

I would expect it to be done *before* calling `lsInner.getChanges()` so that the meta data in the database contains this information?

Reply:
this actually makes me think that maybe we should remove the explicit calls to `updateLastModified` from the codebase and instead have ledger state do it:
* every time "commit" is called
* every time one of the getChanges is called
* we should crash if `updateLastModified` would modify entries when calling getLive/dead entries (something horrible happened)

it's fairly cheap to do and that way we don't introduce yet another way for people to shoot themselves in the foot.

Also, we should guard against calling `updateLastModified` after commit - as we're modifying ledger entries